### PR TITLE
Store votes and update governance tab

### PR DESCRIPTION
**Pull request by panleone**
The scope of this PR is saving governance votes.

### CODE REVIEW

**Review by Duddino — COMMENTED** [review]
image
Functionally it works, but I would change the color a bit (A more pronounced gradient and remove the purple outline

**Review by JSKitty — CHANGES_REQUESTED** [review]
Personally, I think pulling votes from the network would be better (even if it's more network intensive and RPC-reliant), reasons being:
- A Masternode "imported" to MPW may have already voted, it's votes would not display.
- Unnecessary storage use (which may make compact "full data" exports harder to achieve).
- Potential for bugs: without a lot of futureproofing, the addition of multi-MN could make it difficult for MPW to associate votes to a given masternode.

This can be done using the `getbudgetvotes <"name of proposal">` RPC, where MPW would iteratively call the RPC for each Governance proposal, and search for votes matching it's own masternode ID (`tx` + `vin`), ideally caching this per-session in memory to avoid spamming the RPC and also save on bandwidth.

**Comment by panleone** [comment]
> 
I disagree, you have to think at how the system will scale when we will have more users. With your idea, the system will scale really bad and we will get literally flooded by useless RPC calls:
On average a proposal has ≈500/600 votes let's say that on an average we have 10 proposals  at the same time, we will have to ask for 5000/6000 votes for each user for each time he clicks on the governance tab, this  just scalea bad and must be avoided.

"Unnecessary storage use", I absolutely disagree we are talking about storing a few bytes.


"Potential of bugs",  I also disagree here the system can be really easily generalized to more than one masternode

However I agree on the first point: if a users imports an existing masternode we have to sync his votes, so I will add that RPC call ONLY one time when the masternode  is imported and then store with the current system, what you think?

**Comment by Duddino** [comment]
>>
> However I agree on the first point: if a users imports an existing masternode we have to sync his votes, so I will add that RPC call ONLY one time when the masternode is imported and then store with the current system, what you think?

If we do this, there's gonna issues if the user votes on another instance/wallet, since it would only sync the first time.
Maybe we can add a filter option to the RPC node? If we generalize it, it could be used for other things too

**Comment by JSKitty** [comment]
> ... the system will scale really bad and we will get literally flooded by useless RPC calls...

I don't see it as flooding, if there's 10 proposals: each user would sync the 10 proposal votes **one time** (not every click of the Governance tab) and cache the results so it never has to call again for that session, so really it's only a measure of "session restarts" rather than users.

>  ... I absolutely disagree we are talking about storing a few bytes.

The current implementation holds the data permanently, so even if you voted on a proposal 6 months ago (which is now far ended) MPW still holds the data, if you vote 100s of times, then MPW will be storing useless old data forever, unless we store timestamps and "cleanup" routinely.

> ... I will add that RPC call ONLY one time when the masternode is imported and then store with the current system, what you think?

That's not a bad middleground, but I still think it's not really a finished solution (if the user has multiple MPW or PIVX Core clients, and votes between them, then the MPW instances will not display the same vote data, they will desync).

IMO, we could either:
- Use a cache-heavy network sync: per session, so on the FIRST click of the Governance tab, votes are synchronised and cached, this cache is lost on session restart.
- We add a manual "Sync Governance" button somewhere, so that the user themselves have the option to "kick" a cache refresh of the Governance data from the network, without automated spam.

**Comment by panleone** [comment]
> I don't see it as flooding, if there's 10 proposals: each user would sync the 10 proposal votes **one time** (not every click of the Governance tab) and cache the results so it never has to call again for that session, so really it's only a measure of "session restarts" rather than users.
> 
Still too much imo, each vote is about 0.31KB and we have to multiply this by ~5000 per user

> The current implementation holds the data permanently, so even if you voted on a proposal 6 months ago (which is now far ended) MPW still holds the data, if you vote 100s of times, then MPW will be storing useless old data forever, unless we store timestamps and "cleanup" routinely.
>
Ok but this is easily fixable 

 
> That's not a bad middleground, but I still think it's not really a finished solution (if the user has multiple MPW or PIVX Core clients, and votes between them, then the MPW instances will not display the same vote data, they will desync).
> 
> IMO, we could either:
> 
> * Use a cache-heavy network sync: per session, so on the FIRST click of the Governance tab, votes are synchronised and cached, this cache is lost on session restart.
> * We add a manual "Sync Governance" button somewhere, so that the user themselves have the option to "kick" a cache refresh of the Governance data from the network, without automated spam.

I think I will just go for Duddino's idea and modify the PIVX Node Wallet with a function that filter only your vote, I think is the best possibility

**Comment by JSKitty** [comment]
I think an RPC filter will be fine for this case - I currently tend to avoid adding "middleware" between MPW and PIVX Core with custom functionality, but in this case it's a huge data saver for a basic filter, and we still can have a network vote sync. :+1:

**Comment by BreadJS** [comment]
> image Functionally it works, but I would change the color a bit (A more pronounced gradient and remove the purple outline

The color change will happen as soon as I will work on that page!

**Comment by panleone** [comment]
Updated using filters! Now votes are stored only for the local session in order to have instantaneous button update.
I'll leave the color as they are now since Bread is going to revisit the page anyways (?)

**Review by Duddino — COMMENTED** [review]
Network filter works, loading the page now only takes 10 kB.
I still don't like the button color though

**Comment by Duddino** [comment]
image
image
I'm by no means a color expert, but these colors are a bit better. Sent the patch to discord

**Comment by panleone** [comment]
rebased on master in order to test with ledger

**Review by vampyren — APPROVED** [review]
Works great with ledger and MN but i can still vote with network "off", not sure if its a bug or not?

**Review by JSKitty — APPROVED** [review]
**utACK** b465ca1d29e3eaca088b6fe536b68f62e096623a, on behalf of @vampyren

File: scripts/masternode.js [hunk 1]
```diff
@@ -349,7 +353,47 @@ export default class Masternode {
...
+        //See if you already voted the proposal in the current session
+        const index = Masternode.sessionVotes.findIndex(
+            ([vHash]) => vHash === hash
+        );
+        if (index !== -1) {
+            //Found it! return the vote
+            return Masternode.sessionVotes[index][1];
+        }
+        //Haven't voted yet, fetch the result from Duddino's node
+        const filter = `.[]%20%7C%20select(.mnId=="${this.collateralTxId}-${this.outidx}")`;
```
Comment (Duddino) [inline scripts/masternode.js:371]: Leave this as plain text and encode it with `encodeURI` for readability

File: scripts/masternode.js [hunk 2]
```diff
@@ -349,7 +353,47 @@ export default class Masternode {
...
+        const url = `${cNode.url}/getbudgetvotes?params=${proposalName}&filter=${filter}`;
+        const text = await (await fetch(url)).text();
```
Comment (Duddino) [inline scripts/masternode.js:373]:
```suggestion
        const vote = await (await fetch(url)).json();
```

File: scripts/masternode.js [hunk 3]
```diff
@@ -349,7 +353,47 @@ export default class Masternode {
...
+        try {
+            return JSON.parse(text).Vote === 'YES' ? 1 : 2;
```
Comment (Duddino) [inline scripts/masternode.js:375]:
```suggestion
            return vote.Vote === 'YES' ? 1 : 2;
```

**Review by JSKitty — APPROVED** [review]
**utACK**, performant and clean code, Quality Control team response is good.

**Pull request merged**